spark/src/test/java/org/apache/iceberg/spark/source/TestSparkSchema.java

```diff
@@ -0,0 +1,214 @@
+      optional(2, "data", Types.StringType.get())
+  );
+  private static SparkSession spark = null;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @BeforeClass
+  public static void startSpark() {
+    TestSparkSchema.spark = SparkSession.builder().master("local[2]").getOrCreate();
+  }
+
+  @AfterClass
+  public static void stopSpark() {
+    SparkSession currentSpark = TestSparkSchema.spark;
+    TestSparkSchema.spark = null;
+    currentSpark.stop();
+  }
+
+  @Test
+  public void testSparkReadSchemaIsHonored() throws IOException {
+    String tableLocation = temp.newFolder("iceberg-table").toString();
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.unpartitioned();
+    tables.create(SCHEMA, spec, null, tableLocation);
+
+    List<SimpleRecord> expectedRecords = Lists.newArrayList(
+        new SimpleRecord(1, "a")
+    );
+    Dataset<Row> originalDf = spark.createDataFrame(expectedRecords, SimpleRecord.class);
+    originalDf.select("id", "data").write()
+        .format("iceberg")
+        .mode("append")
+        .save(tableLocation);
+
+    StructType sparkReadSchema =
+        new StructType(
+            new StructField[] {
+                new StructField("id", DataTypes.IntegerType, true, Metadata.empty())
+            }
+        );
+
+    Dataset<Row> resultDf = spark.read()
+        .schema(sparkReadSchema)
+        .format("iceberg")
+        .load(tableLocation);
+
+    Row[] results = (Row[]) resultDf.collect();
+
+    Assert.assertEquals("Result size matches", 1, results.length);
+    Assert.assertEquals("Row length matches with sparkReadSchema", 1, results[0].length());
+    Assert.assertEquals("Row content matches data", 1, results[0].getInt(0));
+  }
+
+  @Test
+  public void testEmptyRowsIfSparkReadSchemaIsOff() throws IOException {
+    String tableLocation = temp.newFolder("iceberg-table").toString();
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.unpartitioned();
+    tables.create(SCHEMA, spec, null, tableLocation);
+
+    List<SimpleRecord> expectedRecords = Lists.newArrayList(
+        new SimpleRecord(1, "a")
+    );
+    Dataset<Row> originalDf = spark.createDataFrame(expectedRecords, SimpleRecord.class);
+    originalDf.select("id", "data").write()
+        .format("iceberg")
+        .mode("append")
+        .save(tableLocation);
+
+    StructType sparkReadSchema =
+        new StructType(
+            new StructField[] {
+                new StructField("idd", DataTypes.IntegerType, true, Metadata.empty()) // wrong field name
+            }
+        );
+
+    Dataset<Row> resultDf = spark.read()
+        .schema(sparkReadSchema)
+        .format("iceberg")
+        .load(tableLocation);
+
+    Row[] results = (Row[]) resultDf.collect();
+
+    Assert.assertEquals("Result size matches", 1, results.length);
+    Assert.assertEquals("Row length matches with sparkReadSchema", 0, results[0].length());
```

Comment:
I'm not sure that this is correct. If the user requests a field that is not present in the table then this produces a result set that ignores that column. Shouldn't this result in an error because the table doesn't contain the column?

An assumption of the projection code path (pruneColumns) is that the requested columns will be a subset of the columns in the table. We should probably validate that.

Reply:
This is like the last test case, `testFailSparkReadSchemaCombinedWithProjectionWhenSchemaDoesNotContainProjection`. That case fails.

Reply:
Done.